components/menu/menu-item-mixin.js

```diff
@@ -0,0 +1,146 @@
+			hasChildView: { type: Boolean, attribute: 'has-child-view' },
+			hidden: { type: Boolean, reflect: true },
+			last: { type: String, reflect: true }, // set by d2l-menu
+			role: { type: String, reflect: true },
+			tabindex: { type: Number, reflect: true, attribute: 'tab-index' },
+			text: String
+		};
+	}
+
+	constructor() {
+		super();
+		this.__keyCodes = {
+			ENTER: 13,
+			LEFT: 37,
+			RIGHT: 39,
+			SPACE: 32
+		};
+		this.__children = null;
+		this.role = 'menuitem';
+		this.tabindex = -1;
+	}
+
+	firstUpdated() {
+		super.firstUpdated();
+		this.__initializeItem();
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+
+		changedProperties.forEach((oldValue, propName) => {
+			if (propName === 'hidden') {
+				this._onHidden();
+			}
+		});
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		this.addEventListener('blur', this.__onBlur);
+		this.addEventListener('click', this.__onClick);
+		this.addEventListener('d2l-hierarchical-view-hide-complete', this.__onHideComplete);
+		this.addEventListener('dom-change', this.__onDomChange);
+		this.addEventListener('focus', this.__onFocus);
+		this.addEventListener('keydown', this.__onKeyDown);
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		this.removeEventListener('blur', this.__onBlur);
+		this.removeEventListener('click', this.__onClick);
+		this.removeEventListener('d2l-hierarchical-view-hide-complete', this.__onHideComplete);
+		this.removeEventListener('dom-change', this.__onDomChange);
+		this.removeEventListener('focus', this.__onFocus);
+		this.removeEventListener('keydown', this.__onKeyDown);
+	}
+
+	__initializeItem() {
+		const children = Array.from(this.children);
+		if (children && children.length > 0 && children[0].tagName === 'TEMPLATE') {
+			return;
+		}
+
+		for (let i = 0; i < children.length; i++) {
+			if (children[i].tagName !== 'TEMPLATE' && children[i].tagName !== 'DOM-IF') {
+				this.hasChildView = true;
+				this.__children = children;
+				this.setAttribute('aria-haspopup', true);
+				this.__children[0].label = this.text;
+				break;
+			}
+		}
+		if (this._initialize) {
+			this._initialize();
+		}
+	}
+
+	__action() {
+
+		if (this.disabled) {
+			return;
+		}
+
+		if (this.__children && this.__children.length > 0 && this.__children[0].isHierarchicalView) {
+			// assumption: single, focusable child view
+			this.__children[0].show();
+		} else {
+			this.dispatchEvent(new CustomEvent('d2l-menu-item-select', { bubbles: true, composed: true }));
+		}
+	}
+
+	__onBlur() {
+		this.dispatchEvent(new CustomEvent('_blur', { bubbles: true, composed: true }));
+	}
+
+	__onClick(e) {
+		e.stopPropagation();
+		this.__action();
+	}
+
+	__onDomChange() {
+		this.__initializeItem();
+	}
+
+	__onFocus() {
+		this.dispatchEvent(new CustomEvent('_focus', { bubbles: true, composed: true }));
+	}
+
+	__onHideComplete(e) {
+		if (this.__children.indexOf(e.target) === -1) {
```

Comment:
This also previously used polymer dom (https://github.com/BrightspaceUI/menu/blob/master/d2l-menu-item-behavior.js#L163)

Reply:
I'm not 100% certain, but I think the purpose of this event handler is to return focus to the "opener" menu-item of a nested menu when drilling out of a nested menu.  If there are multiple levels of nested menus, we only want to return focus to this item if the even originated from a **direct** child menu.  I think this change looks good, but I think there should be a demo that can be used to test this out.  :)

Reply:
Yup that seems to be the scenario that causes this to be needed. I've been testing in the `d2l-menu` demo and that demo contains multiple levels of nested menus which captures this scenario. That should definitely get moved over once menu is moved to lit.